rts/Lua/LuaUnsyncedCtrl.cpp

```diff
@@ -3382,6 +3384,226 @@ int LuaUnsyncedCtrl::SetBuildFacing(lua_State* L)
+			return false;
+		}
+	}
+
+	int unitDefID = lua_toint(L, idx);
+	float3 pos = {lua_tofloat(L, idx+1), lua_tofloat(L, idx+2), lua_tofloat(L, idx+3)};
+	int facing = lua_toint(L, idx+4);
+
+	buildInfo = BuildInfo(unitDefID, pos, facing);
+	return true;
+}
+
+
+static bool ParseBuildInfoTable(
+	lua_State* L,
+	BuildInfo& buildInfo,
+	const char* caller,
+	int idx
+) {
+	if (!lua_istable(L, idx)) {
+		luaL_error(L, "%s(): bad build info type", caller);
+		return false;
+	}
+
+	lua_rawgeti(L, idx, 1);
+	if (!lua_isnumber(L, -1)) {
+		luaL_error(L, "%s(): bad unit def ID", caller);
+		return false;
+	}
+	int unitDefID = lua_toint(L, -1);
+	lua_pop(L, 1);
+
+	float3 pos;
+	for (int i=0; i<3; i++) {
+		lua_rawgeti(L, idx, i+2);
+		if (!lua_isnumber(L, -1)) {
+			luaL_error(L, "%s(): bad pos", caller);
+			return false;
+		}
+		pos[i] = lua_tofloat(L, -1);
+		lua_pop(L, 1);
+	}
+
+	lua_rawgeti(L, idx, 5);
+	if (!lua_isnumber(L, -1)) {
+		luaL_error(L, "%s(): bad facing", caller);
+		return false;
+	}
+	int facing = lua_toint(L, -1);
+	lua_pop(L, 1);
+
+	buildInfo = BuildInfo(unitDefID, pos, facing);
+	return true;
+}
+
+
+static bool ParseBuildInfoArray(
+	lua_State* L,
+	std::vector<BuildInfo>& buildInfos,
+	const char* caller,
+	int idx
+) {
+	if (!lua_istable(L, idx)) {
+		luaL_error(L, "%s(): error parsing build info array", caller);
+		return false;
+	}
+
+	for (lua_pushnil(L); lua_next(L, idx) != 0; lua_pop(L, 1)) {
+		buildInfos.emplace_back();
+		if (!ParseBuildInfoTable(L, buildInfos.back(), caller, -1)) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+
+static bool ParseBuildSquareOptions(
+	lua_State* L,
+	LuaBuildSquareOptions& opts,
+	const char* caller,
+	const int idx
+) {
+	if (lua_istable(L, idx)) {
+		for (lua_pushnil(L); lua_next(L, idx) != 0; lua_pop(L, 1)) {
+			// "key" = value (table format of CommandNotify)
+			// ignore the "coded" key; not a boolean value
+			if (lua_israwstring(L, -2)) {
+				if (lua_isboolean(L, -1)) {
+					const bool value = lua_toboolean(L, -1);
+					switch (hashString(lua_tostring(L, -2))) {
+						case hashString("unbuildable"):
+							opts.unbuildable = value;
+							break;
+					}
+				} else if (lua_istable(L, -1)) {
+					float color[4];
+					LuaUtils::ParseFloatArray(L, -1, &color[0], 4);
+					switch (hashString(lua_tostring(L, -2))) {
+						case hashString("defaultColor"):
+							opts.defaultColor = &color[0];
+							break;
+						case hashString("buildableColor"):
+							opts.buildableColor = &color[0];
+							break;
+						case hashString("unbuildableColor"):
+							opts.unbuildableColor = &color[0];
+							break;
+						case hashString("featureColor"):
+							opts.featureColor = &color[0];
+							break;
+						case hashString("illegalColor"):
+							opts.illegalColor = &color[0];
+							break;
+					}
+				}
+			}
+		}
+	} else if (!lua_isnil(L, idx)) {
+		luaL_error(L, "%s(): bad options-argument type", caller);
+		return false;
+	}
+	return true;
+}
+
+
+/***
+ *
```

Comment:
A description on where to use the function would be good. Something like "best used in wupget:DrawFoo since that is the phase engine draws it natively, also works in DrawBaz and DrawQuux"

Reply:
done